File: byteback-test/test-scala-2/src/main/scala/byteback/test/exceptions/PotentialNullDereference.scala
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} --nct -o %t.bpl
 */

package byteback.test.exceptions;

import byteback.annotations.Contract._;
import byteback.annotations.Special._;
import byteback.annotations.Operator._;
import byteback.annotations.Operator.{eq => equal};

class A @Return () {

		@Return
		def method(): Unit = {}

}

class PotentialNullDereference {

		@Return
		def freshTarget(): Unit = {
				var a: A = new A()
				a.method()
		}

		@Return
		def passedTarget(a: A): Unit = {
				if (a != null) {
						a.method()
				}
		}

}

/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 5 verified, 0 errors
 */

File: byteback-test/test-kotlin-1.8.0/src/main/kotlin/byteback/test/exceptions/PotentialNullDereference.kt
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} --nct -o %t.bpl
 */

package byteback.test.exceptions

import byteback.annotations.Contract.*
import byteback.annotations.Operator.*

class A @Return constructor() {

		@Return
		fun method() {}

}

class PotentialNullDereference {

		@Return
		fun freshTarget() {
				val a: A = A()
				a.method()
		}

		@Return
		fun passedTarget(a: A?) {
				if (a != null) {
						a.method()
				}
		}

		@Return
		fun passedArrayTarget(a: Array<A>?) {
				if (a != null) {
						val a_l: Int = a.size;
				}
		}

}

/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 6 verified, 0 errors
 */